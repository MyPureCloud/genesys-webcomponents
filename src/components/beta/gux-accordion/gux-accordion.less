gux-accordion-beta {
  -custom-noop: noop;
}
