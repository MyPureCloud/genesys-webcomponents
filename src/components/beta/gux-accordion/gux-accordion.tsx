import { Component, Element, h } from '@stencil/core';

import { trackComponent } from '../../../usage-tracking';

@Component({
  styleUrl: 'gux-accordion.less',
  tag: 'gux-accordion-beta',
  shadow: true
})
export class GuxAccordionBeta {
  @Element()
  root: HTMLElement;

  componentWillLoad() {
    trackComponent(this.root);
  }

  render() {
    return <slot></slot>;
  }
}
