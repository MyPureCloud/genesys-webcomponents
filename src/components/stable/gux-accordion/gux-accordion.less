@import (reference) '../../../style/color.less';

// Variables part

gux-accordion {
  .gux-accordion {
    padding: 0;
    margin: 0;

    .section {
      .header {
        box-sizing: border-box;
        height: 40px;

        button {
          font-family: inherit;
          font-size: inherit;
          line-height: 40px;
          font-weight: bold;
          border: none;
          background: none;
          display: block;
          cursor: pointer;
          width: 100%;
          text-align: left;
          padding: 0;

          .heading-text {
            padding: 0 0 0 16px;
          }
        }

        .toggle-arrow {
          float: right;
          padding: 0 12px;
          height: 39px;
          display: flex;
          align-items: center;

          gux-icon {
            width: 16px;
            height: 16px;
          }
        }
      }

      .content {
        box-sizing: border-box;
        display: none;
        padding: 16px;
      }

      &.opened {
        .header {
          .toggle-arrow {
            transform: scaleY(-1);
          }
        }

        .content {
          display: block;
        }
      }
    }
  }
}

// Theming
// Light
.gux-accordion-light-theme {
  .gux-accordion {
    background-color: @gux-background;

    .section {
      color: @gux-charcoal-grey;

      .header {
        border-top: 1px solid @gux-border;

        &:hover {
          .toggle-arrow {
            color: @gux-genesys-dark-blue;
          }
        }
      }

      &.opened {
        .header {
          border-bottom: 1px solid @gux-border;
        }
      }

      &:last-of-type {
        .header {
          border-bottom: 1px solid @gux-border;
        }

        .content {
          border-bottom: 1px solid @gux-border;
        }
      }
    }
  }
}

.gux-light-theme {
  gux-accordion {
    .gux-accordion-light-theme();
  }
}

gux-accordion.gux-light-theme {
  .gux-accordion-light-theme();
}
// Dark
.gux-accordion-dark-theme {
  ul.gux-accordion {
    background-color: @gux-background-dark;

    .section {
      color: @gux-type-dark;

      .header {
        border-top: 1px solid @gux-grey-global;

        .toggle-arrow {
          color: @gux-type-dark;
        }

        &:hover {
          .toggle-arrow {
            color: @gux-genesys-blue;
          }
        }
      }

      &.opened {
        .header {
          border-bottom: 1px solid @gux-grey-global;
        }
      }

      &:last-of-type {
        .header {
          border-bottom: 1px solid @gux-grey-global;
        }

        .content {
          border-bottom: 1px solid @gux-grey-global;
        }
      }
    }
  }
}

.gux-dark-theme {
  gux-accordion {
    .gux-accordion-dark-theme();
  }
}

gux-accordion.gux-dark-theme {
  .gux-accordion-dark-theme();
}
// Default
gux-accordion {
  .gux-accordion-light-theme();
}
