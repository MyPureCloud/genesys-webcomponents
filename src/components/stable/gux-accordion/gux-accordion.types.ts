export type GuxAccordionArrowPosition = 'beside-text' | 'default';

export interface IGuxAccordionSection {
  slotName: string;
  ref: HTMLElement;
}
