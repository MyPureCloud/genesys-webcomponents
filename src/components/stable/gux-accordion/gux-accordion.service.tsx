import { IGuxAccordionSection } from './gux-accordion.types';

function getSectionByName(
  slotName: string,
  sections: IGuxAccordionSection[]
): IGuxAccordionSection {
  return sections.find(section => section.slotName === slotName);
}

function getPreviousSection(
  slot: string,
  sections: IGuxAccordionSection[]
): HTMLElement {
  const currentIndex = sections.findIndex(section => section.slotName === slot);

  if (currentIndex <= 0) {
    return sections[sections.length - 1].ref;
  }

  return sections[currentIndex - 1].ref;
}

function getNextSection(
  slot: string,
  sections: IGuxAccordionSection[]
): HTMLElement {
  const currentIndex = sections.findIndex(section => section.slotName === slot);

  if (currentIndex >= sections.length - 1) {
    return sections[0].ref;
  }

  return sections[currentIndex + 1].ref;
}

function getFirstSection(sections: IGuxAccordionSection[]): HTMLElement {
  return sections[0].ref;
}

function getLastSection(sections: IGuxAccordionSection[]): HTMLElement {
  return sections[this.sections.length - 1].ref;
}

function getToggleButton(section: HTMLElement): HTMLElement {
  return section.querySelector('.gux-header-button');
}

export function getSections(root: HTMLElement): IGuxAccordionSection[] {
  const children = Array.from(root.children) as HTMLElement[];
  const sections: IGuxAccordionSection[] = [];

  children.forEach(element => {
    const slotName = element.getAttribute('slot');

    element.hidden = !Boolean(slotName);

    if (slotName) {
      sections.push({
        slotName,
        ref: null
      });
    }
  });

  return sections;
}

export function modifyClassList(
  slotName: string,
  modification: 'add' | 'remove' | 'toggle',
  sections: IGuxAccordionSection[]
): void {
  const section = getSectionByName(slotName, sections);

  section?.ref.classList[modification]('gux-opened');
}

export function onKeyboardNavigation(
  event: KeyboardEvent,
  slotName: string,
  sections: IGuxAccordionSection[]
): void {
  let newSection: HTMLElement;

  switch (event.key) {
    case 'ArrowUp':
      newSection = getPreviousSection(slotName, sections);
      break;
    case 'ArrowDown':
      newSection = getNextSection(slotName, sections);
      break;
    case 'End':
      newSection = getLastSection(sections);
      break;
    case 'Home':
      newSection = getFirstSection(sections);
      break;
  }

  if (newSection) {
    getToggleButton(newSection).focus();
  }
}
