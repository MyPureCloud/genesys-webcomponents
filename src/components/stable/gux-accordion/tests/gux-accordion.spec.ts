import { newSpecPage } from '@stencil/core/testing';
import { GuxAccordion } from '../gux-accordion';

const components = [GuxAccordion];
const language = 'en';

describe('gux-accordion', () => {
  it('should render component as expected', async () => {
    const html = `
      <gux-accordion heading-level="2">
        <div slot="First Section">
          <span>I'm a span in a div.</span>
          <button>I'm the button.</button>
        </div>
        <p slot="Second Section">I'm a p.</p>
        <span slot="Third Section">I'm a span.</span>
        <span
          slot="Forth Section has a really really long title to see what it looks like when the title overflows"
          >I'm a span.</span
        >
        <h1>I'm an h1, but i'm not a slot.</h1>
      </gux-accordion>
    `;
    const page = await newSpecPage({ components, html, language });

    expect(page.root).toMatchSnapshot();
  });
});
