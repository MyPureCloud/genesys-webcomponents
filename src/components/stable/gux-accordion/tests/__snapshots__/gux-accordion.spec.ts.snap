// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gux-accordion should render component as expected 1`] = `
<gux-accordion heading-level="2">
  <!---->
  <h1 hidden="">
    I'm an h1, but i'm not a slot.
  </h1>
  <div class="gux-accordion">
    <section class="gux-section">
      <div aria-level="2" aria-role="heading" class="gux-header">
        <button class="gux-header-button" type="button">
          <div class="gux-text">
            First Section
          </div>
          <div class="gux-spacer"></div>
          <div class="gux-toggle-arrow">
            <gux-icon decorative="" icon-name="chevron-small-down"></gux-icon>
          </div>
        </button>
      </div>
      <div class="gux-content">
        <div slot="First Section">
          <span>
            I'm a span in a div.
          </span>
          <button>
            I'm the button.
          </button>
        </div>
      </div>
    </section>
    <section class="gux-section">
      <div aria-level="2" aria-role="heading" class="gux-header">
        <button class="gux-header-button" type="button">
          <div class="gux-text">
            Second Section
          </div>
          <div class="gux-spacer"></div>
          <div class="gux-toggle-arrow">
            <gux-icon decorative="" icon-name="chevron-small-down"></gux-icon>
          </div>
        </button>
      </div>
      <div class="gux-content">
        <p slot="Second Section">
          I'm a p.
        </p>
      </div>
    </section>
    <section class="gux-section">
      <div aria-level="2" aria-role="heading" class="gux-header">
        <button class="gux-header-button" type="button">
          <div class="gux-text">
            Third Section
          </div>
          <div class="gux-spacer"></div>
          <div class="gux-toggle-arrow">
            <gux-icon decorative="" icon-name="chevron-small-down"></gux-icon>
          </div>
        </button>
      </div>
      <div class="gux-content">
        <span slot="Third Section">
          I'm a span.
        </span>
      </div>
    </section>
    <section class="gux-section">
      <div aria-level="2" aria-role="heading" class="gux-header">
        <button class="gux-header-button" type="button">
          <div class="gux-text">
            Forth Section has a really really long title to see what it looks like when the title overflows
          </div>
          <div class="gux-spacer"></div>
          <div class="gux-toggle-arrow">
            <gux-icon decorative="" icon-name="chevron-small-down"></gux-icon>
          </div>
        </button>
      </div>
      <div class="gux-content">
        <span slot="Forth Section has a really really long title to see what it looks like when the title overflows">
          I'm a span.
        </span>
      </div>
    </section>
  </div>
</gux-accordion>
`;
